core/src/main/java/org/apache/calcite/sql/SqlTableIdentifierWithID.java

```diff
@@ -0,0 +1,314 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.util.SqlVisitor;
+import org.apache.calcite.sql.validate.SqlMonotonicity;
+import org.apache.calcite.sql.validate.SqlTableIdentifierWithIDQualified;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+import org.apache.calcite.tools.ValidationException;
+import org.apache.calcite.util.Litmus;
+import org.apache.calcite.util.Util;
+
+import com.google.common.collect.ImmutableList;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.dataflow.qual.Pure;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A <code>SqlTableIdentifierWithID</code> is an identifier, possibly compound.
+ * This differs from SqlIdentifier because it only refers to tables that will have
+ * an extra column not found in their type, __bodo_row_id. We use this identifier
+ * to uniquely identify particular tables of interest, for example the target table
+ * in MergeInto.
+ */
+public class SqlTableIdentifierWithID extends SqlNode {
+  /**
+   * Array of the components of this compound identifier.
+   **
+   */
+  public ImmutableList<String> names;
+
+  /**
+   * This identifier's collation (if any).
+   */
+  final @Nullable SqlCollation collation;
+
+  /**
+   * A list of the positions of the components of compound identifiers.
+   */
+  protected @Nullable ImmutableList<SqlParserPos> componentPositions;
+
+  /** Create a SqlTableIdentifierWithID from identifier components. This is used in case
+   * we make a copy. This is mostly copied from SQLIdentifier.
+   */
+  public SqlTableIdentifierWithID(
+      List<String> names,
+      @Nullable SqlCollation collation,
+      SqlParserPos pos,
+      @Nullable List<SqlParserPos> componentPositions) {
+    super(pos);
+    this.names = ImmutableList.copyOf(names);
+    this.collation = collation;
+    this.componentPositions = componentPositions == null ? null
+        : ImmutableList.copyOf(componentPositions);
+    for (String name : names) {
+      assert name != null;
+      // We don't support wildcards
+      assert !name.equals("");
+    }
+  }
+
+  /**
+   * Create the identifier from an existing SqlIdentifier.
+   * @param id Original SQL identifier for a table.
+   */
+  public SqlTableIdentifierWithID(SqlIdentifier id) {
+    super(id.pos);
+    this.names = ImmutableList.copyOf(id.names);
+    this.collation = id.collation;
+    this.componentPositions = componentPositions == null ? null
+        : ImmutableList.copyOf(componentPositions);
+    for (String name : names) {
+      assert name != null;
+      // We don't support wildcards
+      assert !name.equals("");
+    }
+
+  }
+
+  @Override public SqlKind getKind() {
```

Comment:
I'm assuming everything past this point is a copy paste from SqlIdentifier, since I didn't see any comments to the contrary. If there's something specific you want me to look at, lmk.

Reply:
These are basically copy paste except that I raise an exception if we find a function.